packages/playwright-test/src/reporters/merge.ts

```diff
@@ -153,3 +156,61 @@ async function sortedShardFiles(dir: string) {
   const files = await fs.promises.readdir(dir);
   return files.filter(file => file.endsWith('.jsonl')).sort();
 }
+
+class ProjectNamePatcher {
+  constructor(private _projectNameSuffix: string) {
+  }
+
+  patchEvents(events: JsonEvent[]) {
+    if (!this._projectNameSuffix)
+      return;
+    for (const event of events) {
+      const { method, params } = event;
+      switch (method) {
+        case 'onBegin':
+          this._onBegin(params.config, params.projects);
+          continue;
+        case 'onTestBegin':
+        case 'onStepBegin':
+        case 'onStepEnd':
+        case 'onStdIO':
+          params.testId = this._mapTestId(params.testId);
+          continue;
+        case 'onTestEnd':
+          params.test.testId = this._mapTestId(params.test.testId);
+          continue;
+      }
+    }
+  }
+
+  private _onBegin(config: JsonConfig, projects: JsonProject[]) {
+    for (const project of projects)
+      project.name += this._projectNameSuffix;
+    this._updateProjectIds(projects);
+    for (const project of projects)
+      project.suites.forEach(suite => this._updateTestIds(suite));
+  }
+
+  private _updateProjectIds(projects: JsonProject[]) {
+    const usedNames = new Set<string>();
+    for (const p of projects) {
+      for (let i = 0; i < projects.length; ++i) {
+        const candidate = p.name + (i ? i : '');
+        if (usedNames.has(candidate))
+          continue;
+        p.id = candidate;
+        usedNames.add(candidate);
+        break;
+      }
+    }
+  }
+
+  private _updateTestIds(suite: JsonSuite) {
+    suite.tests.forEach(test => test.testId = this._mapTestId(test.testId));
+    suite.suites.forEach(suite => this._updateTestIds(suite));
+  }
+
+  private _mapTestId(testId: string): string {
+    return testId + '-' + this._projectNameSuffix;
```

Comment:
The only thing I don't really like is that we are patching test ids. These are supposed to be "history ids", so it seems fragile that changing the suffix changes test ids. However, I don't see a way around it.